src/box/lua/console.lua

```diff
@@ -649,15 +667,27 @@ local function local_read(self)
             break
         end
         if delim == "" then
-            local lang = self.language
-            if not lang or lang == 'lua' then
-                -- stop once a complete Lua statement is entered
-                if local_check_lua(buf) then
+            local continuation_on
+            if self.remote ~= nil then
+                continuation_on = self.remote.continuation_on
+            else
+                continuation_on = self.continuation_on
+            end
+            local lang = self.language or 'lua'
+            -- Continue reading if Lua input seems incomplete according to 
+            -- internal heuristic.
+            if lang ~= 'lua' or local_check_lua(buf) then
+                -- Interpret continuation_symbol as a line continuation marker.
+                if not continuation_on or
```

Comment:
Well, I agree with @Totktonada here: AFAICS, this is the only condition that need to be added; all other code can be left untouched (correct me if I'm wrong). I propose to implement patch in another way, instead of introducing such a complex waterfall of conditions for a single `break`. If I did understand your code right, the next changes do the same:
```diff
diff --git a/src/box/lua/console.lua b/src/box/lua/console.lua
index bcdfd79f0..83e17a64c 100644
--- a/src/box/lua/console.lua
+++ b/src/box/lua/console.lua
@@ -649,12 +649,17 @@ local function local_read(self)
             break
         end
         if delim == "" then
-            local lang = box.session.language
-            if not lang or lang == 'lua' then
+            local lang = self.language or 'lua'
+            if lang == 'lua' then
                 -- stop once a complete Lua statement is entered
                 if local_check_lua(buf) then
                     break
                 end
+                if continuation_on and line:endswith(continuation_symbol) then
+                    buf = buf:sub(1, -1 - #continuation_symbol)
+                else
+                    break
+                end
             else
                 break
             end
```

**Of course** with a bunch of the relevant comments for posterity ;)

---

P.S. Honestly, I've misread `lang ~= 'lua' or local_check_lua(buf)` a couple of times and thought that `not` is missing somewhere...

Reply:
With `continue` it looks muck better:
```diff
diff --git a/src/box/lua/console.lua b/src/box/lua/console.lua
index 9df668b42..11b2d2b61 100644
--- a/src/box/lua/console.lua
+++ b/src/box/lua/console.lua
@@ -677,24 +677,22 @@ local function local_read(self)
                 continuation_on = self.continuation_on
             end
             local lang = self.language or 'lua'
+            -- Interpret continuation_symbol as a line continuation marker.
+            if continuation_on and line:endswith(continuation_symbol) then
+                buf = buf:sub(1, -1 - #continuation_symbol)
+                goto continue
+            end
             -- Continue reading if Lua input seems incomplete according to 
             -- internal heuristic.
-            if lang ~= 'lua' or local_check_lua(buf) then
-                -- Interpret continuation_symbol as a line continuation marker.
-                if not continuation_on or
-                        not line:endswith(continuation_symbol) then
-                    break
-                end
-                -- Continue reading when input ends with line continuation
-                -- marker and is a complete Lua statement or in some other
-                -- language.
-                buf = buf:sub(1, -1 - #continuation_symbol)
+            if lang == 'lua' and not local_check_lua(buf) then
+                goto continue
             end
-            -- Continue reading.
+            break
         elseif #buf >= #delim and buf:sub(#buf - #delim + 1) == delim then
             buf = buf:sub(0, #buf - #delim)
             break
         end
+        ::continue::
         buf = buf.."\n"
         prompt = string.rep(' ', #self.prompt)
     end
```

Reply:
Agreed, now it definitely looks better.

Reply:
Ouch, I resolved Igor's comment. It was unintentional, reopened back.

Reply:
Yeah, definitely better. Thanks for the fixes!